sw/device/silicon_creator/mask_rom/bootstrap.c

```diff
@@ -0,0 +1,321 @@
+  if (tail_byte_count > 0) {
+    size_t padding_byte_count =
+        FLASH_CTRL_PARAM_BYTES_PER_WORD - tail_byte_count;
+    for (size_t i = 0; i < padding_byte_count; ++i) {
+      data[byte_count++] = 0xff;
+    }
+  }
+  size_t word_count = byte_count / sizeof(uint32_t);
+
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolTrue,
+      .erase = kHardenedBoolFalse,
+  });
+  rom_error_t error = flash_ctrl_data_write(addr, word_count, data);
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolFalse,
+      .erase = kHardenedBoolFalse,
+  });
+  return error;
+}
+
+/**
+ * Bootstrap state 1: Wait for an erase command and erase the data
+ * partition.
+ *
+ * This function erases both data banks of the flash regardless of the type of
+ * the erase command (CHIP_ERASE or SECTOR_ERASE).
+ *
+ * @param state Bootstrap state.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_handle_erase(bootstrap_state_t *state) {
+  HARDENED_CHECK_EQ(*state, kBootstrapStateErase);
+
+  spi_device_cmd_t cmd;
+  spi_device_cmd_get(&cmd);
+  rom_error_t error = kErrorUnknown;
+  switch (cmd.opcode) {
+    case kSpiDeviceOpcodeChipErase:
+    case kSpiDeviceOpcodeSectorErase:
+      error = bootstrap_chip_erase();
+      HARDENED_RETURN_IF_ERROR(error);
+      *state = kBootstrapStateEraseVerify;
+      break;
+    default:
+      // Ignore any other command, e.g. PAGE_PROGRAM, RESET.
+      error = kErrorOk;
+  }
+
+  // Note: We clear WIP and WEN bits in `bootstrap_handle_erase_verify()` after
+  // checking that both data banks have been erased.
+  return error;
+}
+
+/**
+ * Bootstrap state 2: Verify that all data banks have been erased.
+ *
+ * This function also clears the WIP and WEN bits of the flash status register.
+ *
+ * @param state Bootstrap state.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_handle_erase_verify(bootstrap_state_t *state) {
+  HARDENED_CHECK_EQ(*state, kBootstrapStateEraseVerify);
+
+  rom_error_t err_0 = flash_ctrl_data_erase_verify(0, kFlashCtrlEraseTypeBank);
+  rom_error_t err_1 = flash_ctrl_data_erase_verify(
+      FLASH_CTRL_PARAM_BYTES_PER_BANK, kFlashCtrlEraseTypeBank);
+  HARDENED_RETURN_IF_ERROR(err_0);
+  HARDENED_RETURN_IF_ERROR(err_1);
+
+  *state = kBootstrapStateProgram;
+  spi_device_flash_status_clear();
+  return err_0;
+}
+
+/**
+ * Bootstrap state 3: (Erase/)Program loop.
+ *
+ * @param state Bootstrap state.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_handle_program(bootstrap_state_t *state) {
+  static_assert(alignof(spi_device_cmd_t) >= sizeof(uint32_t) &&
+                    offsetof(spi_device_cmd_t, payload) >= sizeof(uint32_t),
+                "Payload must be word aligned.");
+
+  HARDENED_CHECK_EQ(*state, kBootstrapStateProgram);
+
+  spi_device_cmd_t cmd;
+  spi_device_cmd_get(&cmd);
```

Comment:
Am I missing the part where we spin until we get an event? It looks like we're spinning and assuming there is always a command available. Did I miss something?

Reply:
`spi_device_cmd_get()` is blocking. I called it from individual handlers as opposed to `bootstrap()` because `bootstrap_handle_erase_verify()` does not handle a spi flash command.

Reply:
Ahh, that explains it--I just wasn't familiar with the API. Thanks! :)

Reply:
Are we enabling the watchdog in the mask_rom as a safety precaution?  If so, does `spi_device_cmd_get` periodically pet the dog to prevent a reboot?

Reply:
Yes, watchdog is enabled in mask_rom but no, `spi_device_cmd_get()` does not periodically pet it. I'm planning to disable the watchdog before calling `bootstrap()` in `mask_rom.c`. WDYT?

Reply:
I'm not sure if it makes much difference, however:

There are some operations in `flash_ctrl.c` which busy-loop waiting for the flash controller status.  Leaving the watchdog enabled protects us from those loops getting stuck or being subverted (e.g. `wait_for_done` waits on a single bit).

Reply:
I'll keep the watchdog out of bootstrap for now for the sake of simplicity. PLMK if you think we should capture this in an issue to revisit later.

Reply:
SGTM, please create an issue to track this.

Reply:
Created #12387